src/curl_xml.c

```diff
@@ -461,11 +465,66 @@ static int cx_handle_instance_xpath(xmlXPathContextPtr xpath_ctx, /* {{{ */
    * somewhere inside this structure. */
   xmlXPathFreeObject(instance_node_obj);
 
+  /* Part 2, handle PluginInstanceFrom */
+  instance_node_obj = NULL;
+  instance_node = NULL;
+
+  /* plugin_instance_from has to be an xpath expression */
+  if (xpath->plugin_instance_from != NULL) {
+    int tmp_size;
+
+    instance_node_obj = cx_evaluate_xpath(xpath_ctx,
```

Comment:
This same sequence of

*   `cx_evaluate_xpath()`
*   `if (num <= 0)`
*   `if (num > 1)`
*   `if (!cx_if_not_text_node())`

Is repeated in `cx_handle_instance_xpath()` and `cx_handle_single_value_xpath()`. I'd recommend to unify this in a function with approximately the following prototype:

```c
int cx_evaluate_xpath_strbuf(xmlXPathContextPtr xpath_ctx,
                             xmlChar *expr,
                             char *out_buf, size_t out_buf_size);
```